qiskit_experiments/curve_analysis/curve_analysis.py

```diff
@@ -148,6 +151,118 @@ def model_names(self) -> List[str]:
         """Return model names."""
         return [getattr(m, "_name", f"model-{i}") for i, m in enumerate(self._models)]
 
+    def set_options(self, **fields):
+        """Set the analysis options for :meth:`run` method.
+
+        Args:
+            fields: The fields to update the options
+
+        Raises:
+            KeyError: When removed option ``curve_fitter`` is set.
+        """
+        if fields.get("plot_residuals") and not self.options.get("plot_residuals"):
+            # checking there are no subplots for the figure to prevent collision in subplot indices.
+            if self.plotter.options.get("subplots") != (1, 1):
+                warnings.warn(
+                    "Residuals plotting is currently supported for analysis with 1 subplot.",
+                    UserWarning,
+                    stacklevel=2,
+                )
+                fields["plot_residuals"] = False
+            else:
+                self._add_residuals_plot_config()
+        if not fields.get("plot_residuals", True) and self.options.get("plot_residuals"):
+            self._remove_residuals_plot_config()
+
+        super().set_options(**fields)
+
+    def _add_residuals_plot_config(self):
+        """Configure plotter options for residuals plot."""
+        # check we have model to fit into
+        residual_plot_y_axis_size = 3
+        if self.models:
+            # Cache figure options.
+            self._plot_config_cache["figure_options"] = {}
+            self._plot_config_cache["figure_options"]["ylabel"] = self.plotter.figure_options.get(
+                "ylabel"
+            )
+            self._plot_config_cache["figure_options"]["series_params"] = deepcopy(
+                self.plotter.figure_options.get("series_params")
+            )
+            self._plot_config_cache["figure_options"]["sharey"] = self.plotter.figure_options.get(
+                "sharey"
+            )
+
+            self.plotter.set_figure_options(
+                ylabel=[
+                    self.plotter.figure_options.get("ylabel", ""),
+                    "Residuals",
+                ],
+            )
+
+            model_names = self.model_names()
+            series_params = self.plotter.figure_options["series_params"]
+            for model_name in model_names:
+                if series_params.get(model_name, {}):
+                    series_params[model_name]["canvas"] = 0
+                else:
+                    series_params[model_name] = {"canvas": 0}
+                series_params[model_name + "_residuals"] = series_params[model_name].copy()
```

Comment:
How do you decide when to copy, and if so, whether to copy or deep-copy?

Reply:
I used deep-copy for the caching. If I use copy on a `dict` (like  `figure_options["series_params"]` and then I change it, it will changed the cached value.